### PR TITLE
Diagram metadata

### PR DESCRIPTION
**Pull request by amolenaar**
### PR Type
What kind of change does this PR introduce?

- [ ] Bug fix
- [x] Feature
- [ ] Chore (refactoring, formatting, local variables, other cleanup)
- [ ] Documentation content changes

### What is the current behavior?

Issue Number: #803

### What is the new behavior?

Add a diagram item that can provide diagram meta data.

- [ ] Should this diagram item be backed by a model element, or is this data really diagram specific?
- [x] Add Property page to edit fields in metadata block
- [x] Hide unused fields?

### CODE REVIEW

**Comment by danyeaw** [comment]
How do I add metadata to a diagram?

**Comment by danyeaw** [comment]
How do I add metadata to a diagram?

**Comment by danyeaw** [comment]
How do I add metadata to a diagram?

**Comment by amolenaar** [comment]
I forgot: it still needs a proper icon.

**Comment by amolenaar** [comment]
@danyeaw @vanillajonathan Added an icon.

image

NB. All fields are saved in the diagram item. There's no model element with "shared" data.

File: gaphor/diagram/general/metadata.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+
+class MetadataItem(ElementPresentation):
+    createdBy: attribute[str] = attribute("createdBy", str, "")
+    description: attribute[str] = attribute("description", str, "")
+    website: attribute[str] = attribute("website", str, "")
+    revision: attribute[str] = attribute("revision", str, "")
+    license: attribute[str] = attribute("license", str, "")
+    createdAt: attribute[str] = attribute("createdAt", str, "")
+    updatedAt: attribute[str] = attribute("updatedAt", str, "")
```
Comment (danyeaw) [inline gaphor/diagram/general/metadata.py:28]: I would expect that Created/Updated At would be for a location, I suggest Created On and Updated On would be more clear for a date.

File: gaphor/diagram/general/metadata.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+        self.shape = Box(
+            *(
+                [
+                    Box(
+                        *(
+                            [
+                                Box(
+                                    Text(
+                                        text=gettext("created by:"),
```
Comment (danyeaw) [inline gaphor/diagram/general/metadata.py:66]: Kind of nit, but I think the metadata would look better if it was in Title Case with each work capitalized.

**Review by danyeaw — APPROVED** [review]
This looks fantastic, I really like how it turned out! It is simple to add metadata to a diagram, and I think it was the right approach to save it to the diagram instead of creating a separate element.

I have two minor comments about naming.

File: gaphor/diagram/general/metadata.py (hunk 2 above)
Comment (vanillajonathan) [reply gaphor/diagram/general/metadata.py:66]: Yeah, or all caps but in a small font like it was in my proposal.
https://github.com/gaphor/gaphor/issues/803#issuecomment-986680585

File: gaphor/diagram/general/metadata.py (hunk 1 above)
Comment (amolenaar) [reply gaphor/diagram/general/metadata.py]: fixed

File: gaphor/diagram/general/metadata.py (hunk 2 above)
Comment (amolenaar) [reply gaphor/diagram/general/metadata.py]: Let's stick with Title Case indeed. It's more consistent with the rest of the app.

**Comment by amolenaar** [comment]
> Looks good, maybe the Gtk.Entry for the license could have proposals for some SPDX license identifiers such as `CC-PDDC`. `CC0-1.0`, `CC-BY-NC-4.0`, `CC-BY-NC-SA-4.0`, `CC-BY-SA-4.0`, `GFDL-1.3-or-later`.

I'd like to postpone this until we're rid of the GTK3 code. GTK4 does not have a nice editable dropdown by itself (`GtkComboBoxText` is deprecated).

**Comment by vanillajonathan** [comment]
> > Looks good, maybe the Gtk.Entry for the license could have proposals for some SPDX license identifiers such as `CC-PDDC`. `CC0-1.0`, `CC-BY-NC-4.0`, `CC-BY-NC-SA-4.0`, `CC-BY-SA-4.0`, `GFDL-1.3-or-later`.
> 
> I'd like to postpone this until we get rid of the GTK3 code. GTK4 does not have a nice editable dropdown by itself (`GtkComboBoxText` is deprecated).

GTK 3 have [`EntryCompletion`](https://docs.gtk.org/gtk3/class.EntryCompletion.html) for the `Gtk.Entry` widget. GTK 4 also have [`EntryCompletion`](https://docs.gtk.org/gtk3/class.EntryCompletion.html)but it is deprecated since 4.1.0.

https://discourse.gnome.org/t/replacement-for-entrycompletion/13505

**Pull request merged**